sdk/storage_blobs/src/clients/container_client.rs

```diff
@@ -15,30 +19,83 @@ use azure_storage::{
+#[derive(Clone, Debug)]
+pub struct ContainerClientBuilder {
+    storage_credentials: Option<StorageCredentials>,
+    storage_account: String,
+    container: String,
+    storage_options: StorageOptions,
 }
 
-impl AsContainerClient for StorageClient {
-    fn container_client(&self, container_name: impl Into<String>) -> ContainerClient {
-        ContainerClient::new(self.clone(), container_name.into())
+impl ContainerClientBuilder {
+    // TODO: I actually think it would be best to have the credentails as part of this signature, which would make them "required"
+    // Then rather than having builder methods do so all the credential conversions as part of the Client we just add the functionality
+    // StorageCredentials itself.
+    #[must_use]
+    pub fn new(account: impl Into<String>, container: impl Into<String>) -> Self {
+        let storage_account = account.into();
+        let container = container.into();
+        Self {
+            storage_credentials: None,
+            storage_account,
+            container,
+            storage_options: StorageOptions::new(),
+        }
+    }
+
+    #[must_use]
+    pub fn credentials(mut self, storage_credentials: impl Into<StorageCredentials>) -> Self {
+        self.storage_credentials = Some(storage_credentials.into());
+        self
+    }
+
+    #[must_use]
+    pub fn retry(mut self, retry: impl Into<azure_core::RetryOptions>) -> Self {
+        self.storage_options.options = self.storage_options.options.retry(retry);
+        self
+    }
+
+    #[must_use]
+    pub fn transport(mut self, transport: impl Into<azure_core::TransportOptions>) -> Self {
+        self.storage_options.options = self.storage_options.options.transport(transport);
+        self
+    }
+
+    #[must_use]
+    pub fn timeout(mut self, timeout: impl Into<TimeoutPolicy>) -> Self {
+        let timeout = timeout.into();
+        self.storage_options.timeout_policy = timeout;
+        self
+    }
+
+    pub fn build(self) -> ContainerClient {
+        // TODO: Make this an error?
+        let storage_credentials = self.storage_credentials.unwrap();
+        let pipeline = new_pipeline_from_options(self.storage_options, storage_credentials.clone());
+        let url = get_endpoint_uri(None, &self.storage_account, "blob")
+            .unwrap()
+            .join(&self.container)
+            .unwrap();
+        ContainerClient {
+            storage_account: self.storage_account,
+            storage_credentials,
+            container_name: self.container,
+            url,
+            pipeline,
+        }
     }
 }
 
 #[derive(Debug, Clone)]
 pub struct ContainerClient {
-    storage_client: StorageClient,
+    storage_account: String,
+    storage_credentials: StorageCredentials,
     container_name: String,
+    url: Url,
+    pipeline: Pipeline,
 }
 
 impl ContainerClient {
-    pub(crate) fn new(storage_client: StorageClient, container_name: String) -> Self {
```

Comment:
Do we want to *require* the user to use the builder to construct the client. This is not what we did in Cosmos. I think we should have `ContainerClient::new` which takes an account, container name, credentials and options.